File: soccer/gameplay2/skills/line_kick.py
import skills._kick
import behavior
import constants
import robocup
import enum


# lines up with the ball and the target, then drives up and kicks
# this differs from PivotKick which gets the ball first, then aims
class LineKick(skills._kick._Kick):

    # tuneable constants
    DriveAroundDist = 0.15
    ChargeThresh = 0.1
    EscapeChargeThresh = 0.1
    SetupBallAvoid = 0.08
    AccelBias = 0.2
    FacingThresh = 10   # angle in degrees
    MaxChargeSpeed = 1.5
    BallProjectTime = 0.4
    DoneStateThresh = 0.11
    LandOnTarget = False


    class State(enum.Enum):
        setup = 1
        charge = 2


    def __init__(self):
        super().__init__()
        self.target_point = constants.Field.TheirGoalSegment.center()

        for state in LineKick.State:
            self.add_state(state, behavior.Behavior.State.running)

        self.add_transition(behavior.Behavior.State.start,
            LineKick.State.setup,
            lambda: True,
            'immediately')



    # where we're trying to kick
    # Default: the center of their goal
    @property
    def target_point(self):
        return self._target_point
    @target_point.setter
    def target_point(self, value):
        self._target_point = value


    def recalculate(self):
        self._target_line = robocup.Line(main.ball().pos, self.target_point)

        bot_dir = robocup.Point.direction(self.robot.angle * constants.DegreesToRadians)


    def execute_running(self):
        self.recalculate()


    def execute_setup(self):
        move_goal = main.ball().pos - self._target_line.delta().normalized() * (LineKick.DriveAroundDist + constants.Robot.Radius)

        left_field_edge = robocup.Segment(robocup.Point(-constants.Field.Width / 2.0, 0),
            robocup.Point(-constants.Field.Width / 2.0, constants.Field.Length))
        right_field_edge = robocup.Segment(robocup.Point(constants.Field.Width / 2.0, 0),
            robocup.Point(constants.Field.Width / 2.0, constants.Field.Length))

        # handle the case when the ball is near the field's edge
        field_edge_thresh = 0.3
        behind_line = robocup.Segment(main.ball().pos - self._target_line.delta().normalized() * LineKick.DriveAroundDist,
            main.ball().pos - self._target_line.delta().normalized())
        main.system_state().draw_line(behind_line, constants.Colors.Blue, "LineKick")
        for edge in [left_field_edge, right_field_edge]:
            if edge.near_point(main.ball().pos, field_edge_thresh):
                intersection = behind_line.line_intersection(edge)
                if intersection != None:
                    move_goal = intersection

        self.robot.set_avoid_ball_radius(LineKick.SetupBallAvoid)
        self.robot.move_to(move_goal)
        self.robot.face(self.robot.pos + (self.target_point - main.ball().pos))
        self.robot.unkick()


    def execute_charge(self):
        main.system_state().draw_line(robocup.Line(self.robot.pos, self.target_point), constants.Colors.White, "LineKick")
        main.system_state().draw_line(robocup.Line(main.ball().pos, self.target_point), constants.Colors.White, "LineKick")

        # drive directly into the ball
        ball2target = (self.target_point - main.ball().pos).normalized()
        robot2ball = (main.ball().pos, self.robot.pos).normalized()
        speed = min(self.robot.vel.mag() + LineKick.AccelBias, MaxChargeSpeed)
        self.robot.set_world_vel(robot2ball.normalized() * speed)
        
        self.robot.face(self.target_point)

        if self.use_chipper:
            self.robot.chip(self.chip_power)
        else:
            self.kick(self.kick_power)

File: old/gameplay-old/behaviors/LineKick.cpp
#include "LineKick.hpp"
#include <stdio.h>

using namespace std;
using namespace Geometry2d;

namespace Gameplay
{
	namespace Behaviors
	{
		REGISTER_CONFIGURABLE(LineKick)
	}
}

ConfigDouble *Gameplay::Behaviors::LineKick::_drive_around_dist;
ConfigDouble *Gameplay::Behaviors::LineKick::_setup_to_charge_thresh;
ConfigDouble *Gameplay::Behaviors::LineKick::_escape_charge_thresh;
ConfigDouble *Gameplay::Behaviors::LineKick::_setup_ball_avoid;
ConfigDouble *Gameplay::Behaviors::LineKick::_accel_bias;
ConfigDouble *Gameplay::Behaviors::LineKick::_facing_thresh;
ConfigDouble *Gameplay::Behaviors::LineKick::_max_speed;
ConfigDouble *Gameplay::Behaviors::LineKick::_proj_time;
ConfigDouble *Gameplay::Behaviors::LineKick::_dampening;
ConfigDouble *Gameplay::Behaviors::LineKick::_done_thresh;
ConfigBool *Gameplay::Behaviors::LineKick::_land_on_target;



void Gameplay::Behaviors::LineKick::createConfiguration(Configuration *cfg)
{
	_drive_around_dist = new ConfigDouble(cfg, "LineKick/Drive Around Dist", 0.25);
	_setup_to_charge_thresh = new ConfigDouble(cfg, "LineKick/Charge Thresh", 0.1);
	_escape_charge_thresh = new ConfigDouble(cfg, "LineKick/Escape Charge Thresh", 0.1);
	_setup_ball_avoid = new ConfigDouble(cfg, "LineKick/Setup Ball Avoid", Ball_Radius * 2.0);
	_accel_bias = new ConfigDouble(cfg, "LineKick/Accel Bias", 0.1);
	_facing_thresh = new ConfigDouble(cfg, "LineKick/Facing Thresh - Deg", 10);
	_max_speed = new ConfigDouble(cfg, "LineKick/Max Charge Speed", 1.5);
	_proj_time = new ConfigDouble(cfg, "LineKick/Ball Project Time", 0.4);
	_dampening = new ConfigDouble(cfg, "LineKick/Ball Project Dampening", 0.8);
	_done_thresh = new ConfigDouble(cfg, "LineKick/Done State Thresh", 0.11);
	_land_on_target = new ConfigBool(cfg, "LineKick/Land On Target", false);
}

Gameplay::Behaviors::LineKick::LineKick(GameplayModule *gameplay):
    SingleRobotBehavior(gameplay),
    ballClose(false),
    chip_calib(gameplay)
{
	restart();
	target = Geometry2d::Point(0.0, Field_Length);
	enable_kick = true;
}

void Gameplay::Behaviors::LineKick::restart()
{
	_state = State_Setup;
	use_chipper = false;
	kick_power = 255;
	ballClose = false;
	kick_ready = false;
}

bool Gameplay::Behaviors::LineKick::run()
{
	if (!robot || !robot->visible)
	{
		return false;
	}
	// project the ball ahead to handle movement
	double dt = *_proj_time;
	Point ballPos = ball().pos + ball().vel * dt * _dampening->value();  // projecting
//	Point ballPos = ball().pos; // no projecting
	Line targetLine(ballPos, target);
	const Point dir = Point::direction(robot->angle * DegreesToRadians);
	double facing_thresh = cos(*_facing_thresh * DegreesToRadians);
	double facing_err = dir.dot((target - ballPos).normalized());
	

	if(ballPos.distTo(robot->pos) <= *_done_thresh)
	{
		ballClose = true;
	}

	// State changes
	if (_state == State_Setup)
	{
		if (targetLine.distTo(robot->pos) <= *_setup_to_charge_thresh &&
				targetLine.delta().dot(robot->pos - ballPos) <= -Robot_Radius &&
				facing_err >= facing_thresh &&
				robot->vel.mag() < 0.05)
		{
			if(enable_kick)
			{
				_state = State_Charge;
			}
			kick_ready = true;

		}
		else
		{
			kick_ready = false;
		}

		//if the ball if further away than the back off distance for the setup stage
		if(ballClose && ballPos.distTo(robot->pos) > *_drive_around_dist + Robot_Radius)
		{
			_state = State_Done;
		}
	} else if (_state == State_Charge)
	{
		if (Line(robot->pos, target).distTo(ballPos) > *_escape_charge_thresh)
		{
			// Ball is in a bad place
			_state = State_Setup;
		}

		//if the ball if further away than the back off distance for the setup stage
		if(ballClose && ballPos.distTo(robot->pos) > *_drive_around_dist + Robot_Radius)
		{
			_state = State_Done;
		}
	}
	
	// Driving
	if (_state == State_Setup)
	{
		// Move onto the line containing the ball and the_setup_ball_avoid target
		robot->addText(QString("%1").arg(targetLine.delta().dot(robot->pos - ballPos)));
		Point moveGoal = ballPos - targetLine.delta().normalized() * (*_drive_around_dist + Robot_Radius);

		static const Segment left_field_edge(Point(-Field_Width / 2.0, 0.0), Point(-Field_Width / 2.0, Field_Length));
		static const Segment right_field_edge(Point(Field_Width / 2.0, 0.0), Point(Field_Width / 2.0, Field_Length));

		// Handle edge of field case
		float field_edge_thresh = 0.3;
		Segment behind_line(ballPos - targetLine.delta().normalized() * (*_drive_around_dist),
				ballPos - targetLine.delta().normalized() * 1.0);
		state()->drawLine(behind_line);
		Point intersection;
		if (left_field_edge.nearPoint(ballPos, field_edge_thresh) && behind_line.intersects(left_field_edge, &intersection))   /// kick off left edge of far half fieldlPos, field_edge_thresh) && behind_line.intersects(left_field_edge, &intersection))
		{
			moveGoal = intersection;
		} else if (right_field_edge.nearPoint(ballPos, field_edge_thresh) && behind_line.intersects(right_field_edge, &intersection))
		{
			moveGoal = intersection;
		}

		robot->addText("Setup");
		robot->avoidBallRadius(*_setup_ball_avoid);
		robot->move(moveGoal);

		// face in a direction so that on impact, we aim at goal
		Point delta_facing = target - ballPos;
		robot->face(robot->pos + delta_facing);

		robot->kick(0);

	} else if (_state == State_Charge)
	{
		robot->addText("Charge!");

		state()->drawLine(robot->pos, target, Qt::white);
		state()->drawLine(ballPos, target, Qt::white);
		Point ballToTarget = (target - ballPos).normalized();
		Point robotToBall = (ballPos - robot->pos).normalized();
		Point driveDirection = robotToBall;

		// Drive directly into the ball
		double speed = min(robot->vel.mag() + *_accel_bias, _max_speed->value()); // enough of a bias to force it to accelerate
		





		robot->worldVelocity(driveDirection.normalized() * speed);

		robot->face(target);

		if (use_chipper)
		{
			uint8_t chip_power = kick_power;
			if(*_land_on_target)
				chip_power = chip_calib.chipPowerForDistance((target - ballPos).mag());
			robot->chip(chip_power);
		} else
		{
			robot->kick(kick_power);
		}

	} else {
		robot->addText("Done");
		return false;
	}
	return true;
}
